### PR TITLE
feat: Add publishable key scopes middleware

### PR DESCRIPTION
**Pull request by olivermrbl**
**What**

Add pub key + sales channel middlewares to the store carts API
- Assign sales channel associated with pub key, if sales channel is not passed in request
  - Throw if pub key has multiple associated sales channels
- Throw if sales channel ID in payload is not associated with publishable API key in header

### CODE REVIEW

**Review by fPolic — APPROVED** [review]
LGTM

File: packages/medusa/src/api-v2/utils/middlewares/common/ensure-pub-key-sales-channel-match.ts [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+    if (scopes.sales_channel_ids.length && !channelId) {
+      if (scopes.sales_channel_ids.length > 1) {
+        req.errors.push(
+          `Cannot assign sales channel to cart. The Publishable API Key in the header has multiple associated sales channels. Please provide a sales channel ID in the request body.`
+        )
+      } else {
+        req.validatedBody.sales_channel_id = scopes.sales_channel_ids[0]
+      }
+    }
+  }
```
Comment (adrien2p) [inline packages/medusa/src/api-v2/utils/middlewares/common/ensure-pub-key-sales-channel-match.ts]:
```suggestion
if (scopes.sales_channel_ids.length) {
  const hasSalesChannelId = !!channelId
  const isSalesChannelIdValid = hasSalesChannelId && scopes.sales_channel_ids.includes(channelId)
  
  if (!hasSalesChannelId) {
    req.validatedBody.sales_channel_id = scopes.sales_channel_ids[0]
    return next()
  }
  
  if (scopes.sales_channel_ids.length > 1 && !hasSalesChannelId) {
    req.errors.push(
          `Cannot assign sales channel to cart. The Publishable API Key in the header has multiple associated sales channels. Please provide a sales channel ID in the request body.`
        )
  }
  
  if (!isSalesChannelIdValid) {
    req.errors.push(
        `Sales channel ID in payload ${channelId} is not associated with the Publishable API Key in the header.`
      )
  }
}

next()
```

**Review by adrien2p — APPROVED** [review]
LGTM

File: packages/medusa/src/api-v2/utils/middlewares/common/ensure-pub-key-sales-channel-match.ts (hunk 1 above)
Comment (adrien2p) [reply packages/medusa/src/api-v2/utils/middlewares/common/ensure-pub-key-sales-channel-match.ts]: The above is just a suggestion feel free to take or leave 👍

Comment (olivermrbl) [reply packages/medusa/src/api-v2/utils/middlewares/common/ensure-pub-key-sales-channel-match.ts]: I like it, will update

File: packages/medusa/src/api-v2/utils/middlewares/common/ensure-pub-key-sales-channel-match.ts [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+    if (pubKeySalesChannels.length) {
+      if (channelId && !pubKeySalesChannels.includes(channelId)) {
+        req.errors.push(
+          `Sales channel ID in payload ${channelId} is not associated with the Publishable API Key in the header.`
+        )
+      }
+
+      if (!channelId) {
+        if (pubKeySalesChannels.length > 1) {
```
Comment (sradevski) [inline packages/medusa/src/api-v2/utils/middlewares/common/ensure-pub-key-sales-channel-match.ts:40]: This should not be allowed in the `api-key` module, right? I think it's fine to have the check but in theory it should never happen?

File: packages/medusa/src/api-v2/utils/middlewares/common/ensure-pub-key-sales-channel-match.ts [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+ *         sales channel id is passed in the request.
+ */
+export async function ensurePublishableKeyAndSalesChannelMatch(
+  req: MedusaRequest<StoreCreateCartType> & { publishableApiKeyScopes },
+  res: MedusaResponse,
+  next: NextFunction
+) {
+  const pubKey = req.get("x-publishable-api-key")
+
+  if (pubKey) {
```
Comment (sradevski) [inline packages/medusa/src/api-v2/utils/middlewares/common/ensure-pub-key-sales-channel-match.ts:25]: `if (!pubKey)` so we can return early

File: packages/medusa/src/api-v2/utils/middlewares/common/ensure-pub-key-sales-channel-match.ts [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+    const pubKeySalesChannels =
+      req.publishableApiKeyScopes?.sales_channel_ids ?? []
+    const channelId = req.validatedBody?.sales_channel_id
```
Comment (sradevski) [inline packages/medusa/src/api-v2/utils/middlewares/common/ensure-pub-key-sales-channel-match.ts:28]: Can the `sales_channel_id` also be in the query params?

File: packages/medusa/src/api-v2/utils/middlewares/common/maybe-attach-pub-key-scopes.ts [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+ *         but that id is not in the scope defined by the PK from the header
+ */
+export async function maybeAttachPublishableKeyScopes(
+  req: MedusaRequest & { publishableApiKeyScopes: any },
+  res: MedusaResponse,
+  next: NextFunction
+) {
+  const pubKey = req.get("x-publishable-api-key")
+
+  if (pubKey) {
```
Comment (sradevski) [inline packages/medusa/src/api-v2/utils/middlewares/common/maybe-attach-pub-key-scopes.ts:27]: Same, we can return early here

**Comment by sradevski** [comment]
@olivermrbl A bit late to the party, but left some comments 😄

File: packages/medusa/src/api-v2/utils/middlewares/common/ensure-pub-key-sales-channel-match.ts (hunk 2 above)
Comment (olivermrbl) [reply packages/medusa/src/api-v2/utils/middlewares/common/ensure-pub-key-sales-channel-match.ts:40]: Not 100% sure what you mean. The sales channels for the publishable key are retrieved via the link and attached to the scopes in the other middleware in the PR. These are not from the API key module.

File: packages/medusa/src/api-v2/utils/middlewares/common/ensure-pub-key-sales-channel-match.ts (hunk 3 above)
Comment (olivermrbl) [reply packages/medusa/src/api-v2/utils/middlewares/common/ensure-pub-key-sales-channel-match.ts:25]: We never enter the `if` so we are, effectively, returning early if it is not there. Do you want this for readability?

File: packages/medusa/src/api-v2/utils/middlewares/common/ensure-pub-key-sales-channel-match.ts (hunk 2 above)
Comment (sradevski) [reply packages/medusa/src/api-v2/utils/middlewares/common/ensure-pub-key-sales-channel-match.ts:40]: @olivermrbl Ah sorry I wasn't clear, what I meant is when you associate an api key with a sales channel, we shouldn't allow you to have 2 sales channels assigned to the same publishable key. I incorrectly referenced the API key module, my bad.

File: packages/medusa/src/api-v2/utils/middlewares/common/ensure-pub-key-sales-channel-match.ts (hunk 3 above)
Comment (sradevski) [reply packages/medusa/src/api-v2/utils/middlewares/common/ensure-pub-key-sales-channel-match.ts:25]: @olivermrbl yeah for readability, it reduces the nesting of the code

**Pull request merged**